designs/helper_to_upgrade_projects_by_rescaffolding.md

```diff
@@ -0,0 +1,170 @@
+```
+
+## Open Questions
+
+N/A
+
+## Summary
+
+Therefore, a new command can be designed to load user configs from the [PROJECT config][project-config] file, and run the corresponding kubebuilder subcommands to generate the project based on the new kubebuilder version. Thus, it makes it easier for the users to migrate their operator projects to the new scaffolding.
+
+## Motivation
+
+A common scenario is to upgrade the project based on the newer Kubebuilder. The recommended (straightforward) steps are:
+
+- a) re-scaffold all files from scratch using the upper version/plugins
+- b) copy user-defined source code to the new layout
+
+The proposed command will automate the process at maximum, therefore helping operator authors with minimizing the manual effort.
+
+The main motivation of this proposal is to provide a helper for upgrades and
+make less painful this process. Examples:
+
+- See the discussion [How to regenerate scaffolding?](https://github.com/kubernetes-sigs/kubebuilder/discussions/2864)
+- From [slack channel By Paul Laffitte](https://kubernetes.slack.com/archives/CAR30FCJZ/p1675166014762669)
+
+### Goals
+
+- Help users upgrade their project with the latest changes
+- Help users to re-scaffold projects from scratch based on what was done previously with the tool
+- Make less painful the process to upgrade
+
+### Non-Goals
+
+- Change the default layout or how the KubeBuilder CLI works
+- Deal with customizations or deviations from the proposed layout
+- Be able to perform the project upgrade to the latest changes without human interactions
+- Deal and support external plugins
+- Provides support to [declarative](https://book.kubebuilder.io/plugins/declarative-v1.html) plugin
+  since it is desired and planned to decouple this solution and donate this plugin to its own authors [More info](https://github.com/kubernetes-sigs/kubebuilder/issues/3186)
+- Provide support old version prior we have the Project config (Kubebuilder < 3x) and the go/v2 layout which exist to ensure  a backwards compatibility with legacy layout provide by Kubebuilder 2x
+
+## Proposal
+
+The proposed solution to achieve this goal is to create an alpha command as described
+in the example section above, see:
+
+```shell
+kubebuilder alpha generate \
+    --from=<path of the project file> 
+    --to=<path where the project should be re-scaffold> 
+    --backup=<path-where the current version of the project should be copied as backup>
+    --plugins=<chain of plugins key that can be used to create the layout with init sub-command>
+```
+
+Where:
+
+- from: [Optional] If not informed then, the command would check it in the current directory
+- to: [Optional] If not informed then, it should be the current repository
+- backup: [Optional] If not informed then, the backup by copying the project to a new directory would not be made
+- plugins:  [Optional] If not informed then, it is the same plugin chain available in the layout field
```

Comment:
At the first glance (for someone new) the `from` and `to` flags seem to be `from` and `to` plugins. Thoughts on making it:
```suggestion
kubebuilder alpha generate \
    --input-dir=<path of the project file> 
    --output-dir=<path where the project should be re-scaffold> 
    --backup=<path-where the current version of the project should be copied as backup>
    --plugins=<chain of plugins key that can be used to create the layout with init sub-command>

Where:

- input-dir: [Optional] If not informed then, the command would check it in the current directory. If the `PROJECT` file does not exist it errors. 
- to: [Optional] If not informed then, it should be the current repository.
- backup: [Optional] If not informed then, the backup would be copied to the current directory.
- plugins:  [Required] If not informed then error out.
```

The two major changes I made are:
- backup: [Optional] - If not informed then, the backup would be copied to the current directory.

As an operator author I would still prefer to have a backup created by default in the current directory. If not I would not want to lose the current project. The user can have the option of not doing a back up through `--no-backup` flag. Thoughts on that?

- plugins:  [Required] If not informed then error out
 The  main use case of this command is to re-generate the project with another plugin. If it is not provided, instead of using the same plugin available and creating a copy, can we instead error? If the user would like to create a project with same APIs and plugin as they already have, they can specify it explicitly in the `plugins` field.

Thoughts on this?

Reply:
HI @varshaprasad96,

> backup: [Optional] - If not informed then, the backup would be copied to the current directory.
We might able to have the backup in the same directory if we create the a new dir prefixed with `.` (example my-operator/.backup/my-operator). I think that is a good idea 👍 

> plugins: [Required] If not informed then error out

Why would be required to inform? if not informed that we will use the value from the PROJECT file that we can found in the layout. That allows we scaffold the project for the same plugin but to get the minor changes. Is that make sense?

Reply:
WDYT about we keep `from:` and `to`? It seems very intuitive. Also, we could allow only inform the project path as well to do this operation. (In this case the backup would not be made)

Reply:
@camilamacedo86 `from`, `to` seems generic - its confusing if its the directory from where input project will be taken and re-created or the plugin names which need to be converted. I'd still prefer `input-dir` and `output-dir` to be specific, but we can change it later during the implementation too.